lib/ansible/playbook/role/__init__.py

```diff
@@ -496,7 +496,9 @@ def get_default_vars(self, dep_chain=None):
         dep_chain = [] if dep_chain is None else dep_chain
 
         default_vars = dict()
-        for dep in self.get_all_dependencies():
+        # FIXME combining deps' default_vars can be optimized by moving it
+        #       into _load_role_data so it is done just once
+        for dep in self._dependencies:
             default_vars = combine_vars(default_vars, dep.get_default_vars())
```

Comment:
Can we use the same pattern for defaults and vars, just to make these methods more consistent? For example

```python
        for dep in self.get_all_dependencies():
            default_vars = combine_vars(default_vars, dep._default_vars)
```

or in get_vars maybe
```python
        for dep in self._dependencies:
            # Avoid rerunning dupe deps since they can have vars from previous invocations and they accumulate in deps
            # TODO: re-examine dep loading to see if we are somehow improperly adding the same dep too many times
            if dep not in seen:
                # only take 'exportable' vars from deps
                all_vars = combine_vars(all_vars, dep.get_vars(include_params=False, only_exports=True))
```

Reply:
Both options would be slower, so we would be trading consistency for performance in this case. I am fine with doing:
```python
        for dep in self.get_all_dependencies():
            default_vars = combine_vars(default_vars, dep._default_vars)
```

But the latter option, the `dep.get_vars()` call that only does part of the work depending on the arguments passed is unnecessarily expensive.

Or maybe I can just add comments to explain why it's inconsistent.